<?xml version="1.0"?>
<robot name="roomba" xmlns:xacro="http://ros.org/wiki/xacro">

	<!-- Robot constants and properties -->
	<xacro:include filename="$(find progetto_esame)/urdf/roomba_properties.xacro" />
	<!-- Import Rviz colors -->
	<xacro:include filename="$(find progetto_esame)/urdf/materials.xacro" />
		<!-- Gazebo aspects of the robot -->
	<xacro:include filename="$(find progetto_esame)/urdf/roomba_gazebo.xacro" />
	<!-- Inertia matrices -->
	<xacro:include filename="$(find progetto_esame)/urdf/roomba_macros.xacro" />

	 
	
	<!-- base_footprint is a fictitious link(frame) that is on the ground right below base_link origin, navigation stack depends on this frame -->
	<link name="base_footprint">
	 	<collision>
	      		<origin xyz="0 0 0" rpy="0 0 0" />
	      		<geometry>
	        		<cylinder length="0" radius="0.16495"/>
	     	 	</geometry>
	    	</collision>
	</link>
	
	<!-- base link, basically the body of the Roomba -->
	<link name="base_link">
		<inertial>
	     		<mass value="2" />
	      		<origin xyz="0 0 0.0" />
	      		<inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.5" />
	 	</inertial>
	
	    	<visual>
	      		<origin xyz=" 0 0 0.0308" rpy="0 0 1.5708" />
	      		<geometry>
	        		<mesh filename="package://progetto_esame/meshes/$(arg color)_roomba.dae"/>
	      		</geometry>
	    	</visual>
	
	    	<collision>
	      		<origin xyz="0.0 0.0 0.0358" rpy="0 0 0" />
	      		<geometry>
	        		<cylinder length="0.0711632" radius="0.16495"/>
	     	 	</geometry>
	    	</collision>
	</link>

	<joint name="base_footprint_joint" type="fixed">
		<origin xyz="0 0 0.017" rpy="0 0 0" />        
		<parent link="base_footprint"/>
	 	<child link="base_link" />
	</joint>
	

	<!-- wheels and caster wheel -->

	<xacro:wheel lr="left" tY="-1"/>
	<xacro:wheel lr="right" tY="1"/>

	<link name="caster_wheel_link">
		<inertial>
	      		<origin xyz="0 0 0" />
	      		<mass value="0.01" />
	      		<inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001" />
	    	</inertial>
	
	    	<visual>
	      		<origin xyz="0 0 0" rpy="0 1.5707 1.5707"/>
	      		<geometry>
	        		<!-- <mesh filename="file://$(find progetto_esame)/meshes/caster_wheel.dae"/> -->
				<sphere radius="${casterRadius}" />
	      		</geometry>
			<material name="DarkGray" />
	    	</visual>
	
	    	<collision>
	      		<origin xyz="0 0 0" rpy="0 1.5707 1.5707" />
	      		<geometry>
	        		<sphere radius="${casterRadius}" />
	      		</geometry>
	    	</collision>
	</link>
	
	<joint name="caster_wheel_joint" type="fixed">
	 	<origin xyz="${casterOffsetX} 0 ${casterOffsetZ}" rpy="0 0 0"/>
	    	<parent link="base_link"/>
	    	<child link="caster_wheel_link"/>
	    	<axis xyz="0 1 0"/>
	</joint>

	<!-- Sensor models -->
	<link name="camera_link">
    <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
            <box size="${cameraLength} ${cameraWidth} ${cameraHeight}"/>
        </geometry>
    </collision>

    <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
            <!-- <box size="${cameraLength} ${cameraWidth} ${cameraHeight}"/> -->
			<mesh filename="file://$(find progetto_esame)/meshes/asus_xtion_pro_camera.dae"/>
        </geometry>
        <material name="DarkGray"/>
    </visual>

    <inertial>
        <mass value="${cameraMass}" />
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <xacro:box_inertia m="${cameraMass}" x="${cameraLength}" y="${cameraWidth}" z="${cameraHeight}" />
    </inertial>

  </link>


  <joint name="camera_joint" type="fixed">
    <axis xyz="0 1 0" />
    <origin xyz="${cameraOffsetX} 0 ${cameraOffsetZ}" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="camera_link"/>
  </joint>


  <joint name="hokuyo_joint" type="fixed">
    <axis xyz="0 1 0" />
    <origin xyz="${hokuyoOffsetX} 0 ${hokuyoOffsetZ}" 
            rpy="0 0 3.14159"/>
    <parent link="base_link"/>
    <child link="hokuyo_link"/>
  </joint>


  <!-- Hokuyo Laser -->
  <link name="hokuyo_link">
    <collision>
      <origin xyz="0 0 0" 
              rpy="0 0 0"/>
      <geometry>
        <box size="${hokuyoLength} 
                   ${hokuyoWidth} 
                   ${hokuyoHeight}"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" 
              rpy="0 0 0"/>
      <geometry>
        <mesh filename="file://$(find progetto_esame)/meshes/hokuyo.dae"/>
      </geometry>
    </visual>

    <inertial>
      <mass value="${hokuyoMass}" />
      <origin xyz="0 0 0" 
              rpy="0 0 0"/>
      <xacro:box_inertia 
        m="${hokuyoMass}" 
        x="${hokuyoLength}" 
        y="${hokuyoWidth}" 
        z="${hokuyoHeight}"/>
    </inertial>

  </link>


</robot>
